scripts/builtin/decisionTreePredict.dml

```diff
@@ -0,0 +1,216 @@
+  #  M[5,j]: If j is an internal node: 1 if the feature chosen for j is scale, otherwise the size of the subset of values
+  #    stored in rows 6,7,... if j is categorical
+  #    If j is a leaf node: number of misclassified samples reaching at node j
+  #  M[6:,j]: If j is an internal node: Threshold the example's feature value is compared to is stored at M[6,j] if the feature chosen for j is scale,
+  #     otherwise if the feature chosen for j is categorical rows 6,7,... depict the value subset chosen for j
+  #     If j is a leaf node 1 if j is impure and the number of samples at j > threshold, otherwise 0
+#
+# strategy String  Prediction strategy, can be one of ["GEMM", "TT", "PTT"], referring to "Generic matrix multiplication", 
+#                     "Tree traversal", and "Perfect tree traversal", respectively
+# -------------------------------------------------------------------------------------------
+# OUTPUT   PARAMETERS:
+# ---------------------------------------------------------------------------------------------
+# NAME    TYPE             DEFAULT  MEANING
+# ---------------------------------------------------------------------------------------------
+#	1- Matrix Y containing the predicted labels for X 
+# ---------------------------------------------------------------------------------------------
+
+m_decisionTreePredict = function(
+    Matrix[Double] M,
+    Matrix[Double] X,
+    String strategy
+) return (Matrix[Double] Y) {
+
+  if (strategy == "TT") {
+    Y = predict_TT(M, X)
+  } 
+  else {
+    print ("No such strategy" + strategy)
+    Y = matrix("0", rows=0, cols=0)
+  }
+}
+
+predict_TT = function (
+    Matrix[Double] M,
+    Matrix[Double] X
+) return (Matrix[Double] Y){
+
+  Y = matrix(0, rows=1, cols=nrow(X))
+  n = ncol(M)
+  tree_depth = ceiling(log(7,2)) # max depth of complete binary tree
+  [N_L, N_R, N_F, N_T, N_C] = createNodeTensors(M)
+
+  for (k in 1:nrow(X)){
+    # iterate over every sample in X matrix
+    sample = X[k,]
+    current_node = 1
+    cnt = 1
+    while (cnt < tree_depth){
+      feature_id = as.scalar(N_F[1, current_node])
+      feature = as.scalar(sample[,feature_id]) # select feature from sample data
+      threshold = as.scalar(N_T[1, current_node])
+
+      if (feature < threshold){
+        # move on to left child node
+        next_node = as.scalar(N_L[1, current_node])
+      } else {
+        # move on to right child node
+        next_node = as.scalar(N_R[1, current_node])
+      }
+      current_node = next_node
+      cnt +=1
+    }
+
+    class = M[4, current_node]
+    Y[1, k] = class
+  }
+}
+
+createNodeTensors = function(
+  Matrix[Double] M
+  ) return (
+    Matrix[Double] N_L,
+    Matrix[Double] N_R,
+    Matrix[Double] N_F,
+    Matrix[Double] N_T,
+    Matrix[Double] N_C
+    ){
+      N = M[1,] # all tree nodes
+      I = M[2,] # list of node offsets to their left children
+
+      N_L = matrix(0, rows=1, cols=0)
```

Comment:
i am worried about performance of this implementation of a empty matrix you cbind into.
Since N is known from the start you could allocate the correct size from the beginning.